android/src/com/frostwire/android/gui/activities/ToggleAllSearchEnginesPreference2.java

```diff
@@ -0,0 +1,205 @@
+                    archivePreference = preference;
+                }
+            }
+        }
+
+        // always leave one checked.
+        if (!checked && archivePreference != null) {
+            archivePreference.setOnPreferenceClickListener(null);
+            archivePreference.setChecked(true);
+            archivePreference.setOnPreferenceClickListener(onPreferenceClickListener);
+        }
+    }
+
+    private boolean areAllEnginesChecked() {
+        if (searchEnginePreferences != null) {
+            for (CheckBoxPreference preference : searchEnginePreferences.keySet()) {
+                if (preference != null) {
+                    if (!preference.isChecked()) {
+                        return false;
+                    }
+                }
+            }
+        } else {
+            fixStateAfterNextBind = true;
+        }
+        return true;
+    }
+
+    private boolean isOneEngineNotChecked() {
+        int count = 0;
+        if (searchEnginePreferences != null) {
+            for (CheckBoxPreference preference : searchEnginePreferences.keySet()) {
+                if (preference != null) {
+                    if (!preference.isChecked()) {
+                        count++;
+                    }
+                }
+            }
+            if (count == 1) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isAtLeastOneOtherEngineChecked(CheckBoxPreference checkBoxPreference) {
+        for (CheckBoxPreference preference : searchEnginePreferences.keySet()) {
+            if (preference != null && preference != checkBoxPreference) {
+                if (preference.isChecked()) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public void setSearchEnginePreferences(Map<CheckBoxPreference, SearchEngine> searchEnginePreferences) {
+        this.searchEnginePreferences = searchEnginePreferences;
+    }
+
+    /**
+     * @return True if the child is allowed to change it's state
+     */
+    public boolean requestChildStateChange(CheckBoxPreference checkBoxPreference) {
+        if (isAtLeastOneOtherEngineChecked(checkBoxPreference)) {//can change the child, update state if needed
+            if (checkbox.getHandler() != null) {// only need to animate checkbox if it is visible
+                if (areAllEnginesChecked()) {//turn on the checkbox
+                    checkbox.getHandler().post(delayedCheckboxCheck);
+                } else if (isOneEngineNotChecked()) {//turnoff the check (just unchecked on one of the children)
+                    checkbox.getHandler().post(delayedCheckboxCheck);
+                }
+            } else {
+                checkbox.setChecked(isChecked());
+            }
+            return true;
+        } //can't change the child, don't change anything
+        return false;
+    }
```

Comment:
all these checks seem to eventually do the same thing,

If the checkbox has a handler do `checkbox.setChecked(isChecked());` now.

If all engines are checked, do `checkbox.setChecked(isChecked());` later (aka `delayedCheckboxCheck`)

if one of the engines isn't checked do `checkbox.setChecked(isChecked());` later

Therefore, it seems to me that the `isOneEngineNotChecked()` method isn't needed because you will perform the check anyway.

Perhaps this could be as simple as:

```java
...
if (checkbox.getHandler() != null) {
  checkbox.getHandler().post(delayedCheckboxCheck);
} else {
  checkbox.setChecked(isChecked());
}
return true;
...
```

I'd also rename `isAtLeastOneOtherEngineChecked` to `isOneEngineChecked()`

and would reimplement the whole thing like

```java
public boolean requestChildStateChange(CheckBoxPreference checkBoxPreference) {
    if (!isOneEngineChecked(checkBoxPreference)) {
        // we're done here
        return false;
    }
    if (checkbox.getHandler() != null) {
       // check later
        checkbox.getHandler().post(delayedCheckboxCheck);
    } else {
       // check now
       checkbox.setChecked(isChecked());
    }
    return true;
}
```

Then you can get rid of that `isOneEngineNotChecked()` method and I think it's a lot easier to read.

However, I'd try to re-read the whole thing and see how to simplify, it seems like there are a lot of hacks that will make this hard to maintain. Like what happens if this method returns `true` when invoked above:

```
 if (!selectAll.requestChildStateChange((CheckBoxPreference) cbPreference)) {
                         cbPreference.setChecked(!cbPreference.isChecked());//reverting change we need one
                     }
```

then there's also what happens onBind, not sure if that's there to fix a previous bug, but in my experience, when you start having code that looks like that, at some point later in the future it ends up being removed because it just looks unnecessary.